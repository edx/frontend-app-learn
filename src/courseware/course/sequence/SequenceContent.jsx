import React, { Suspense } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import PageLoading from '../../../PageLoading';
import { useModel } from '../../../model-store';

import messages from './messages';
import Unit from './Unit';

const ContentLock = React.lazy(() => import('./content-lock'));

function SequenceContent({
  gated, intl, courseId, sequenceId, unitId, unitLoadedHandler,
}) {
  const sequence = useModel('sequences', sequenceId);

  if (gated) {
    return (
      <Suspense
        fallback={(
          <PageLoading
            srMessage={intl.formatMessage(messages['learn.loading.content.lock'])}
          />
        )}
      >
        <ContentLock
          courseId={courseId}
          sequenceTitle={sequence.title}
          prereqSectionName={sequence.gatedContent.gatedSectionName}
          prereqId={sequence.gatedContent.prereqId}
        />
      </Suspense>
    );
  }

  if (unitId === null) {
    return (
      <div>
        {intl.formatMessage(messages['learn.sequence.no.content'])}
      </div>
    );
  }

  return (
    <Unit
      courseId={courseId}
      key={unitId}
      id={unitId}
      onLoaded={unitLoadedHandler}
    />
  );
}

SequenceContent.propTypes = {
  gated: PropTypes.bool.isRequired,
  courseId: PropTypes.string.isRequired,
  sequenceId: PropTypes.string.isRequired,
  unitId: PropTypes.string,
  unitLoadedHandler: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

SequenceContent.defaultProps = {
  unitId: null,
};

export default injectIntl(SequenceContent);
