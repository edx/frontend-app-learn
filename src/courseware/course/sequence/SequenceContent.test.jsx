import React from 'react';
import { render, screen } from '../../../test/test-utils';
import SequenceContent from './SequenceContent';

describe('Sequence Content', () => {
  window.scrollTo = jest.fn();
  // HACK: Mock the MutationObserver as it's breaking async testing.
  //  According to StackOverflow it should be fixed in `jest-environment-jsdom` v16,
  //  but upgrading `jest` to v26 didn't fix this problem.
  //  ref: https://stackoverflow.com/questions/61036156/react-typescript-testing-typeerror-mutationobserver-is-not-a-constructor
  global.MutationObserver = class {
    // eslint-disable-next-line no-unused-vars,no-useless-constructor,no-empty-function
    constructor(callback) {}

    disconnect() {}

    // eslint-disable-next-line no-unused-vars
    observe(element, initObject) {}
  };

  const testUnits = [...Array(10).keys()].map(i => String(i + 1));
  const initialState = {
    courseware: {
      sequenceStatus: 'loaded',
      courseStatus: 'loaded',
      courseId: '1',
    },
    models: {
      courses: {
        1: {
          sectionIds: ['1'],
        },
      },
      sections: {
        1: {
          sequenceIds: ['1', '2'],
        },
      },
      sequences: {
        1: {
          unitIds: testUnits,
          showCompletion: true,
          title: 'test-sequence',
          gatedContent: {
            prereqId: '1',
            gatedSectionName: 'test-gated-section',
          },
        },
      },
      units: testUnits.reduce(
        (acc, unitId) => Object.assign(acc, {
          [unitId]: {
            id: unitId,
            contentType: 'other',
            title: unitId,
          },
        }),
        {},
      ),
    },
  };

  const mockData = {
    gated: false,
    courseId: '1',
    sequenceId: '1',
    unitId: '1',
    unitLoadedHandler: () => {},
    intl: {},
  };

  it('displays loading message', () => {
    const { asFragment } = render(<SequenceContent {...mockData} />, { initialState });
    expect(screen.getByText('Loading learning sequence...')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays messages for the locked content', async () => {
    const { asFragment } = render(<SequenceContent {...mockData} gated />, { initialState });
    expect(screen.getByText('Loading locked content messaging...')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();

    expect(await screen.findByText(/content locked/i)).toBeInTheDocument();
    expect(screen.getByText('test-sequence')).toBeInTheDocument();
    expect(screen.queryByText('Loading locked content messaging...')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays message for no content', () => {
    const { asFragment } = render(<SequenceContent {...mockData} unitId={null} />, { initialState });
    expect(screen.getByText('There is no content here.')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
