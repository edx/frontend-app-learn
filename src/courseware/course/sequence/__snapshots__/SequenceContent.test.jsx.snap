// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence Content displays loading message 1`] = `
<DocumentFragment>
  <div
    class="unit"
  >
    <h2
      class="mb-0 h4"
    >
      1
    </h2>
    <button
      aria-disabled="false"
      aria-live="assertive"
      class="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-link px-1 ml-n1 btn-sm"
      type="button"
    >
      <span
        class="d-flex align-items-center justify-content-center"
      >
        <span
          class="pgn__stateful-btn-icon"
        >
          <img
            alt="fa-bookmark"
            class="fa-bookmark"
            data-testid="icon"
          />
        </span>
        <span>
          Bookmark this page
        </span>
      </span>
    </button>
    <div>
      <div
        class="d-flex justify-content-center align-items-center flex-column"
        style="height: 50vh;"
      >
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading learning sequence...
          </span>
        </div>
      </div>
    </div>
    <div
      class="unit-iframe-wrapper"
    >
      <iframe
        allowfullscreen=""
        height="0"
        id="unit-iframe"
        referrerpolicy="origin"
        scrolling="no"
        src="http://localhost:18000/xblock/1?show_title=0&show_bookmark_button=0"
        title="1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sequence Content displays message for no content 1`] = `
<DocumentFragment>
  <div>
    There is no content here.
  </div>
</DocumentFragment>
`;

exports[`Sequence Content displays messages for the locked content 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex justify-content-center align-items-center flex-column"
      style="height: 50vh;"
    >
      <div
        class="spinner-border text-primary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading locked content messaging...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sequence Content displays messages for the locked content 2`] = `
<DocumentFragment>
  <h3>
    <img
      alt="fa-lock"
      class="fa-lock"
      data-testid="icon"
    />
     test-sequence
  </h3>
  <h4>
    Content Locked
  </h4>
  <p>
    You must complete the prerequisite: 'test-gated-section' to access this content.
  </p>
  <p>
    <button
      class="btn btn-primary"
      type="button"
    >
      Go To Prerequisite Section
    </button>
  </p>
</DocumentFragment>
`;

exports[`Unit Navigation displays loading message 1`] = `
<DocumentFragment>
  <div
    class="unit"
  >
    <h2
      class="mb-0 h4"
    >
      1
    </h2>
    <button
      aria-disabled="false"
      aria-live="assertive"
      class="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-link px-1 ml-n1 btn-sm"
      type="button"
    >
      <span
        class="d-flex align-items-center justify-content-center"
      >
        <span
          class="pgn__stateful-btn-icon"
        >
          <img
            alt="fa-bookmark"
            class="fa-bookmark"
            data-testid="icon"
          />
        </span>
        <span>
          Bookmark this page
        </span>
      </span>
    </button>
    <div>
      <div
        class="d-flex justify-content-center align-items-center flex-column"
        style="height: 50vh;"
      >
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading learning sequence...
          </span>
        </div>
      </div>
    </div>
    <div
      class="unit-iframe-wrapper"
    >
      <iframe
        allowfullscreen=""
        height="0"
        id="unit-iframe"
        referrerpolicy="origin"
        scrolling="no"
        src="http://localhost:18000/xblock/1?show_title=0&show_bookmark_button=0"
        title="1"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Unit Navigation displays message for no content 1`] = `
<DocumentFragment>
  <div>
    There is no content here.
  </div>
</DocumentFragment>
`;

exports[`Unit Navigation displays message for the locked content 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex justify-content-center align-items-center flex-column"
      style="height: 50vh;"
    >
      <div
        class="spinner-border text-primary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading locked content messaging...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Unit Navigation displays messages for the locked content 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex justify-content-center align-items-center flex-column"
      style="height: 50vh;"
    >
      <div
        class="spinner-border text-primary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading locked content messaging...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Unit Navigation displays messages for the locked content 2`] = `
<DocumentFragment>
  <h3>
    <img
      alt="fa-lock"
      class="fa-lock"
      data-testid="icon"
    />
     test-sequence
  </h3>
  <h4>
    Content Locked
  </h4>
  <p>
    You must complete the prerequisite: 'test-gated-section' to access this content.
  </p>
  <p>
    <button
      class="btn btn-primary"
      type="button"
    >
      Go To Prerequisite Section
    </button>
  </p>
</DocumentFragment>
`;
